### PR TITLE
sauce-connect: 4.5.4 -> 4.9.1

### PR DESCRIPTION
**Pull request by chasecaleb**
## Description of changes

Latest version on Linux is actually 4.9.2, but latest on Mac is 4.9.1, so only updating to 4.9.1 for consistency between the two.

Also added passthru.updateScript to keep this package up to date in the future, with associated refactoring to make that possible.

P.S. Since this package was significantly out of date, I assume it's effectively unmaintained. I'm willing to become the maintainer of it if that's helpful.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by chasecaleb** [comment]
Proof that I ran `nixpkgs-review` successfully as well as proof that the new version runs (`sc --version`):

```
> nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
$ git worktree add /home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/nixpkgs 12536a0cb51f29a211db96b2d5ac03175e3f6075
Preparing worktree (detached HEAD 12536a0cb51f)
Updating files: 100% (38361/38361), done.
HEAD is now at 12536a0cb51f Merge pull request #272716 from 0xB10C/2023-13-fix-bitcoin-core-darwin-build
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmpjqhd55jy -qaP --xml --out-path --show-trace --no-allow-import-from-derivation
$ git merge --no-commit --no-ff c628cea6f2c022d2107867e0da724af0d2b051c9
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmpjqhd55jy -qaP --xml --out-path --show-trace --no-allow-import-from-derivation --meta
1 package updated:
sauce-connect (4.5.4 → 4.9.1)

$ nix build --nix-path nixpkgs=/home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmpjqhd55jy --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/build.nix
1 package built:
sauce-connect

$ /nix/store/frs3bna4fwzacqfi7mi3zbcryx9pgwi4-nix-2.17.1/bin/nix-shell /home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/shell.nix --nix-path nixpkgs=/home/caleb/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmpjqhd55jy

[nix-shell:~/.cache/nixpkgs-review/rev-c628cea6f2c022d2107867e0da724af0d2b051c9-1]$ sc --version
Sauce Connect 4.9.1, build 2600 55cc68ff
```

**Comment by chasecaleb** [comment]
@offlinehacker FYI since you're listed as maintainer of this package.

File: pkgs/development/tools/sauce-connect/default.nix [hunk 1]
```diff
@@ -2,20 +2,25 @@
...
+    sources = {
+      linux = fetchurl {
+        url = "https://saucelabs.com/downloads/sc-${version}-linux.tar.gz";
+        hash = "sha256-S3vzng6b0giB6Zceaxi62pQOEHysIR/vVQmswkEZ0/M=";
+      };
+      darwin = fetchurl {
+        url = "https://saucelabs.com/downloads/sc-${version}-osx.zip";
+        hash = "sha256-6tJayqo+p7PMz8M651ikHz6tEjGjRIffOqQBchkpW5Q=";
+      };
+    };
```
Comment (wegank) [inline pkgs/development/tools/sauce-connect/default.nix:17]:
```suggestion
    sources = rec {
      x86_64-linux = fetchurl {
        url = "https://saucelabs.com/downloads/sc-${version}-linux.tar.gz";
        hash = "sha256-S3vzng6b0giB6Zceaxi62pQOEHysIR/vVQmswkEZ0/M=";
      };
      x86_64-darwin = fetchurl {
        url = "https://saucelabs.com/downloads/sc-${version}-osx.zip";
        hash = "sha256-6tJayqo+p7PMz8M651ikHz6tEjGjRIffOqQBchkpW5Q=";
      };
      aarch64-darwin = x86_64-darwin;
    };
```

File: pkgs/development/tools/sauce-connect/default.nix [hunk 2]
```diff
@@ -2,20 +2,25 @@
...
+  };
+
+  src =
+    if builtins.elem stdenv.hostPlatform.system lib.platforms.linux
+    then passthru.sources.linux
+    else passthru.sources.darwin;
```
[the remark this answers is not in the document]
Comment (chasecaleb) [reply pkgs/development/tools/sauce-connect/default.nix:23]: Thanks for the review! In response to both this and [your other corresponding suggestion about `passthru.sources`](https://github.com/NixOS/nixpkgs/pull/272774#discussion_r1419816864):

Your suggestions makes sense and is what I was thinking at first, but `meta.platforms` is set to `lib.platforms.linux ++ lib.platforms.darwin` which includes a _lot_ more than just the one linux and two darwin platforms.  So I see a couple options:

1. `passthru.sources` should stay broad like I have it to match `meta.platforms`
2. `meta.platforms` should also be constrained to just these three platforms.
3. Go with your suggestions and don't touch `meta.platforms`. Seems inconsistent/odd, but if that's the right way to do it I'm fine with it :shrug: 

What do you prefer?

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/tools/sauce-connect/default.nix:23]: Set `platforms = builtins.attrNames passthru.sources;`.

**Comment by chasecaleb** [comment]
By the way, OfBorg evaluations won't complete since this package is unfree. I know it's expected for OfBorg not to evaluate it, but it looks like two of the GitHub actions are stuck as "queued" even though they already failed. Is that supposed to happen?

It's these two: 
1. https://github.com/NixOS/nixpkgs/pull/272774/checks?check_run_id=19437545360
2. https://github.com/NixOS/nixpkgs/pull/272774/checks?check_run_id=19437545242

**Comment by chasecaleb** [comment]
Another question: should this be backported since the current version (even in `nixos-unstable`) is too old to work?

See this example, particularly `400: Your Sauce Connect version (4.5.4) is no longer supported`:
```
> NIXPKGS_ALLOW_UNFREE=1 nix shell --impure github:nixos/nixpkgs/nixos-unstable#sauce-connect -c sc -u cchase2 -k "$KEY"
7 Dec 23:13:16 - Sauce Connect 4.5.4, build 4699 f256a9f 
7 Dec 23:13:16 - REST: Using CA certificate bundle /etc/ssl/certs/ca-bundle.crt.
7 Dec 23:13:16 - REST: Using CA certificate verify path /etc/ssl/certs.
7 Dec 23:13:16 - TUNNEL: Using CA certificate bundle /etc/ssl/certs/ca-bundle.crt.
7 Dec 23:13:16 - TUNNEL: Using CA certificate verify path /etc/ssl/certs.
7 Dec 23:13:16 - ***********************************************************
7 Dec 23:13:16 - Warning: open file limit 1024 is too low!
7 Dec 23:13:16 - Sauce Labs recommends setting it to at least 8000
7 Dec 23:13:16 - ***********************************************************
7 Dec 23:13:16 - Starting up; pid 716459
7 Dec 23:13:16 - Command line arguments: sc -u cchase2 -k **** 
7 Dec 23:13:16 - Log file: /tmp/sc.log
7 Dec 23:13:16 - Pid file: /tmp/sc_client.pid
7 Dec 23:13:16 - Timezone: CST GMT offset: -6h
7 Dec 23:13:16 - Using no proxy for connecting to Sauce Labs REST API.
7 Dec 23:13:17 - Error in checkUpdate: couldn't decode JSON document: json: cannot unmarshal string into Go struct field jsonBuild.Build of type int
7 Dec 23:13:17 - Started scproxy on port 34529.
7 Dec 23:13:17 - Please wait for 'you may start your tests' to start your tests.
7 Dec 23:13:17 - error querying from https://saucelabs.com/rest/v1/cchase2/tunnels, error was: 400: Your Sauce Connect version (4.5.4) is no longer supported (must be newer than 4.5.4). Please download the latest version from https://docs.saucelabs.com/secure-connections/sauce-connect/installation (ref: 87d4d64d). HTTP status: 400 Bad Request
7 Dec 23:13:17 - Sauce Connect could not establish a connection.
7 Dec 23:13:17 - Please check your firewall and proxy settings.
7 Dec 23:13:17 - You can also use the sc --doctor to launch Sauce Connect in diagnostic mode.
7 Dec 23:13:17 - Goodbye.
```

**Comment by wegank** [comment]
Services that would fail without up-to-date client software [are backportable](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-acceptable-for-releases).

**Comment by chasecaleb** [comment]
Great, thanks for adding the backport label. Does anything else need to be done on this PR in order to merge it?

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 272774` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/development/tools/sauce-connect/default.nix [hunk 3]
```diff
@@ -2,20 +2,24 @@
...
+    sources = {
+      x86_64-linux = fetchurl {
+        url = "https://saucelabs.com/downloads/sc-${version}-linux.tar.gz";
+        hash = "sha256-S3vzng6b0giB6Zceaxi62pQOEHysIR/vVQmswkEZ0/M=";
+      };
+      x86_64-darwin = fetchurl {
+        url = "https://saucelabs.com/downloads/sc-${version}-osx.zip";
+        hash = "sha256-6tJayqo+p7PMz8M651ikHz6tEjGjRIffOqQBchkpW5Q=";
+      };
+      aarch64-drawin = passthru.sources.x86_64-darwin;
```
[the remark this answers is not in the document]
Comment (chasecaleb) [reply pkgs/development/tools/sauce-connect/default.nix]: :facepalm: Thanks

**Review by wegank — APPROVED** [review]

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 272774` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sauce-connect</li>
  </ul>
</details>

**Pull request merged**